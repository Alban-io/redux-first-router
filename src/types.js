// @flow
export const PREFIX = '@@rudy'
export const prefixType = (type: string) => `${PREFIX}/${type}`

export const UPDATE_HISTORY = prefixType('UPDATE_HISTORY')
export const CALL_HISTORY = prefixType('CALL_HISTORY')
export const NOT_FOUND = prefixType('NOT_FOUND')
export const ADD_ROUTES = prefixType('ADD_ROUTES')
export const CLEAR_CACHE = prefixType('CLEAR_CACHE')
export const CHANGE_BASENAME = prefixType('CHANGE_BASENAME')
