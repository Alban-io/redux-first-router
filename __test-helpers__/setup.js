import createHistory from 'history/createMemoryHistory'
import connectRoutes from '../src/connectRoutes'


export default (path = '/', options = { title: 'title', location: 'location' }) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: '/second/:param',
  }

  const history = createHistory({
    initialEntries: [path],
    initialIndex: 0,
    keyLength: 6,
  })

  return connectRoutes(history, routesMap, options)
}
