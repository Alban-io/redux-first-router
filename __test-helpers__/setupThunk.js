import { createStore, applyMiddleware, compose } from 'redux'
import createHistory from 'history/createMemoryHistory'
import connectTypes from '../src/connectTypes'


export default (path = '/', thunk) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: { path: '/second/:param', thunk },
    THIRD: { path: '/third/:param' },
  }

  const history = createHistory({
    initialEntries: [path],
    initialIndex: 0,
    keyLength: 6,
  })

  const { middleware, enhancer, thunk: ssrThunk, reducer: locationReducer } = connectTypes(history, routesMap)

  const middlewares = applyMiddleware(middleware)

  const reducer = (state = {}, action = {}) => ({
    location: locationReducer(state.location, action),
  })

  const store = createStore(reducer, undefined, compose(enhancer, middlewares))

  return { store, thunk: ssrThunk }
}
