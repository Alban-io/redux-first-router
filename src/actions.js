export const NOT_FOUND = '@@pure-redux-router/NOT_FOUND'

/** PRIMARY/SOLE NAVIGATION ACTION CREATOR
 *  note: on web most of the time you want to use a link component that embeds an `<a />` element
 *  to the DOM for SEO purposes. See our small `<Link />`component:
 *  https://github.com/celebvidy/pure-redux-router-link
*/

export { go } from './connectTypes'
