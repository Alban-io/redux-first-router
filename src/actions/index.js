export { default as redirect } from './redirect'
export { default as notFound } from './notFound'
export { default as addRoutes } from './addRoutes'
export { default as changeBasename } from './changeBasename'

export {
  push,
  replace,
  setState,
  jump,
  back,
  next,
  reset
} from './history'
