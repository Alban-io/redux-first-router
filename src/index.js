export {
  default as connectRoutes,
  go,
  back,
} from './connectRoutes'

export const NOT_FOUND = '@@pure-redux-router/NOT_FOUND'

export { default as actionToPath } from './pure-utils/actionToPath'
export { default as pathToAction } from './pure-utils/pathToAction'

export type {
  RouteString,
  RouteObject,
  Route,
  RoutesMap,
  Routes,
  RouteNames,
  Options,
  Params,
  Payload,
  LocationState,
  Location,
  Meta,
  Action,
  ReceivedAction,
  Listener,
  Listen,
  Push,
  GoBack,
  History,
  HistoryLocation,
  Document,
 } from './flow-types'

