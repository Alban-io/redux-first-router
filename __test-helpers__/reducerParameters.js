import createSmartHistory from '../src/smart-history'

import { createInitialState } from '../src/createLocationReducer'

export default async (type, pathname) => {
  // eslint-disable-line import/prefer-default-export
  const history = createSmartHistory({ initialEntries: ['/first'] })
  history.listen(function() {})
  await history.push(pathname)

  const current = { pathname, type, payload: { param: 'bar' } }
  const prev = { pathname: '/first', type: 'FIRST', payload: {} }
  const routesMap = {
    FIRST: { path: '/first' },
    SECOND: { path: '/second/:param' }
  }

  return {
    type,
    pathname,
    current,
    prev,

    initialState: createInitialState(
      routesMap,
      history
    ),

    routesMap,
    history,

    action: {
      type,
      payload: { param: 'bar' },
      meta: {
        location: {
          current,
          prev,
          kind: 'load',
          history: {
            entries: history.entries.slice(0), // history.entries.map(entry => entry.pathname)
            index: history.index,
            length: history.length
          }
        }
      }
    },

    expectState(state) {
      expect(state.pathname).toEqual(pathname)
      expect(state.type).toEqual(type)
      expect(state.payload).toEqual({ param: 'bar' })
      expect(state.prev).toEqual(prev)
      expect(state.kind).toEqual('load')

      expect(state).toMatchSnapshot()
    }
  }
}
