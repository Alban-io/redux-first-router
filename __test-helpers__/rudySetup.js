import { applyMiddleware, createStore, compose, combineReducers } from 'redux'
import reduxThunk from 'redux-thunk'
import createRouter from '../src/createRouter'
import createScene from '../src/createScene'
import { NOT_FOUND } from '../src/index'

import fakeAsyncWork from '../__test-helpers__/fakeAsyncWork'

export default (path = '/first', options = {}, custom = {}) => {
  const routesMap = {
    ...defaultRoutes,
    ...custom.routesMap
  }

  options.initialEntries = [path]
  options.extra = { arg: 'extra-arg' }

  const { types, actions, routes, exportString } = createScene(routesMap, custom)
  const { middleware, reducer, firstRoute, history } = createRouter(
    routes,
    options
  )

  const title = (state = {}, action = {}) => action.type
  const rootReducer = combineReducers({ title, location: reducer })
  const enhancer = applyMiddleware(middleware, reduxThunk)
  const store = createStore(rootReducer, enhancer)

  return {
    store,
    firstRoute,
    history,
    types,
    actions,
    exportString,
    routes
  }
}

export const defaultRoutes = {
  FIRST: '/first',
  SECOND: '/second',
  THIRD: {
    path: '/third',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    }
  },
  FOURTH: {
    path: '/fourth',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    },
    onComplete: () => {
      return 'onComplete'
    },
    action: (arg) => (req, type) => {
      return { payload: { foo: arg }, type }
    }
  },
  PLAIN: {
    action: (arg) => {
      return { foo: arg }
    }
  },
  [NOT_FOUND]: '/not-found-foo'
}

export const log = store => {
  const state = store.getState().location
  delete state.routesMap
  delete state.hasSSR
  console.log(state)
}
