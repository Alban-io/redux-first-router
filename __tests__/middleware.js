import { setupAll } from '../__test-helpers__/setup'
import { NOT_FOUND } from '../src/types'
import { redirect, notFound } from '../src/actions'

it('dispatches location-aware action, changes address bar + document.title', async () => {
  const { store, history } = await setupAll()

  expect(history.location.pathname).toEqual('/')
  expect(store.getState().location).toMatchSnapshot()

  const params = { param: 'bar' }
  const action = await store.dispatch({ type: 'SECOND', params, state: { foo: 'bar' }, hash: 'foo' })

  store.getState() /*? */

  expect(history.location.pathname).toEqual('/second/bar')
  expect(history.location.state).toEqual({ foo: 'bar' })

  expect(document.title).toEqual('SECOND')

  expect(action).toMatchSnapshot()
  expect(store.getState()).toMatchSnapshot()
})

it('dont double dispatch the same action', async () => {
  const jestNext = jest.fn((next, action) => next(action))
  const additionalMiddleware = store => next => action => jestNext(next, action)
  const { store, history } = await setupAll('/second/bar', undefined, {
    additionalMiddleware
  })

  expect(history.location.pathname).toEqual('/second/bar')
  expect(store.getState().location).toMatchSnapshot()

  const params = { param: 'bar' }
  await store.dispatch({ type: 'SECOND', params })

  console.log(store.getState())
  expect(history.length).toEqual(1)
  expect(jestNext).toHaveBeenCalledTimes(1)
})

it('not matched received action dispatches the action as normal with no changes', async () => {
  const { store, history } = await setupAll('/first')

  expect(history.location.pathname).toEqual('/first')
  expect(store.getState().location).toMatchObject({
    type: 'FIRST',
    pathname: '/first',
    params: {}
  })

  const beforeState = store.getState().location
  const action = await store.dispatch({ type: 'BLA' }) /*? */
  const afterState = store.getState().location /*? */

  expect(action).toEqual({ type: 'BLA' }) // final action returned from middleware is the same as initially dispatched
  expect(history.location.pathname).toEqual('/first') // window.location has not changed because action not matched
  expect(store.getState().location).toMatchObject({
    // location state has not changed because action not matched
    type: 'FIRST',
    pathname: '/first',
    params: {}
  })

  expect(afterState).toEqual(beforeState)
})

it('user dispatches NOT_FOUND and middleware adds missing info to action', async () => {
  const { store } = await setupAll('/first')
  const action = await store.dispatch(notFound({ type: NOT_FOUND }))

  store.getState() /*? $.location */

  expect(action).toMatchSnapshot()
})

it('user dispatches NOT_FOUND redirect and middleware adds missing info to action', async () => {
  const { store } = await setupAll('/first')
  const action = await store.dispatch(redirect({ type: NOT_FOUND }))

  store.getState() /*? $.location */

  expect(action).toMatchSnapshot()
})

it('does nothing if action has error', async () => {
  const { store } = await setupAll('/first')

  const action = {
    error: true,
    type: 'SECOND'
  }

  store.dispatch(action) /*? */
  expect(store.getState().location.type).toEqual('FIRST')
})

it('user redirects', async () => {
  const beforeEnter = ({ action, dispatch }) => {
    if (action.type === 'THIRD') { // also test that a second redirect only results in one entry
      dispatch({ type: 'SECOND', params: { param: 'foo' } })
    }
  }

  const { store, history } = await setupAll('/first', { beforeEnter })
  const action = await store.dispatch(redirect({ type: 'THIRD' }))

  store.getState() /*? $.location */
  expect(store.getState().location.length).toEqual(1)
  expect(store.getState().location.type).toEqual('SECOND')
})
