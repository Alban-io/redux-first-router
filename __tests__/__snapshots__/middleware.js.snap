// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dispatches location-aware action, changes address bar + document.title 1`] = `
Object {
  "basename": "",
  "entries": Array [
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/",
      "search": "",
      "state": Object {},
      "url": "/",
    },
  ],
  "hasSSR": false,
  "hash": "",
  "index": 0,
  "kind": "load",
  "length": 1,
  "params": Object {},
  "pathname": "/",
  "prev": Object {
    "basename": "",
    "entries": Array [],
    "hasSSR": false,
    "hash": "",
    "index": -1,
    "kind": "",
    "length": 0,
    "params": Object {},
    "pathname": "",
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "type": "",
    "url": "",
  },
  "query": Object {},
  "scene": "",
  "search": "",
  "state": Object {},
  "status": 404,
  "type": "@@rudy/NOT_FOUND",
  "url": "/",
}
`;

exports[`dispatches location-aware action, changes address bar + document.title 2`] = `
Object {
  "hash": "foo",
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/",
        "search": "",
        "state": Object {},
        "url": "/",
      },
      Object {
        "hash": "foo",
        "key": "345678",
        "pathname": "/second/bar",
        "search": "",
        "state": Object {
          "foo": "bar",
        },
        "url": "/second/bar#foo",
      },
    ],
    "index": 1,
    "kind": "push",
    "length": 2,
    "pathname": "/second/bar",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "hash": "",
          "key": "345678",
          "pathname": "/",
          "search": "",
          "state": Object {},
          "url": "/",
        },
      ],
      "hasSSR": false,
      "hash": "",
      "index": 0,
      "kind": "load",
      "length": 1,
      "params": Object {},
      "pathname": "/",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 404,
      "type": "@@rudy/NOT_FOUND",
      "url": "/",
    },
    "scene": "",
    "search": "",
    "status": 200,
    "url": "/second/bar#foo",
  },
  "params": Object {
    "param": "bar",
  },
  "query": Object {},
  "state": Object {
    "foo": "bar",
  },
  "type": "SECOND",
}
`;

exports[`dispatches location-aware action, changes address bar + document.title 3`] = `
Object {
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/",
        "search": "",
        "state": Object {},
        "url": "/",
      },
      Object {
        "hash": "foo",
        "key": "345678",
        "pathname": "/second/bar",
        "search": "",
        "state": Object {
          "foo": "bar",
        },
        "url": "/second/bar#foo",
      },
    ],
    "hasSSR": false,
    "hash": "foo",
    "index": 1,
    "kind": "push",
    "length": 2,
    "params": Object {
      "param": "bar",
    },
    "pathname": "/second/bar",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "hash": "",
          "key": "345678",
          "pathname": "/",
          "search": "",
          "state": Object {},
          "url": "/",
        },
      ],
      "hasSSR": false,
      "hash": "",
      "index": 0,
      "kind": "load",
      "length": 1,
      "params": Object {},
      "pathname": "/",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 404,
      "type": "@@rudy/NOT_FOUND",
      "url": "/",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {
      "foo": "bar",
    },
    "status": 200,
    "type": "SECOND",
    "url": "/second/bar#foo",
  },
  "title": "SECOND",
}
`;

exports[`dont double dispatch the same action 1`] = `
Object {
  "basename": "",
  "entries": Array [
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/second/bar",
      "search": "",
      "state": Object {},
      "url": "/second/bar",
    },
  ],
  "hasSSR": false,
  "hash": "",
  "index": 0,
  "kind": "load",
  "length": 1,
  "params": Object {
    "param": "bar",
  },
  "pathname": "/second/bar",
  "prev": Object {
    "basename": "",
    "entries": Array [],
    "hasSSR": false,
    "hash": "",
    "index": -1,
    "kind": "",
    "length": 0,
    "params": Object {},
    "pathname": "",
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "type": "",
    "url": "",
  },
  "query": Object {},
  "scene": "",
  "search": "",
  "state": Object {},
  "status": 200,
  "type": "SECOND",
  "url": "/second/bar",
}
`;

exports[`user dispatches NOT_FOUND and middleware adds missing info to action 1`] = `
Object {
  "hash": "",
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/not-found",
        "search": "",
        "state": Object {},
        "url": "/not-found",
      },
    ],
    "index": 1,
    "kind": "push",
    "length": 2,
    "pathname": "/not-found",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
      ],
      "hasSSR": false,
      "hash": "",
      "index": 0,
      "kind": "load",
      "length": 1,
      "params": Object {},
      "pathname": "/first",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "scene": "",
    "search": "",
    "status": 404,
    "url": "/not-found",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {},
  "type": "@@rudy/NOT_FOUND",
}
`;

exports[`user dispatches NOT_FOUND redirect and middleware adds missing info to action 1`] = `
Object {
  "hash": "",
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/not-found",
        "search": "",
        "state": Object {},
        "url": "/not-found",
      },
    ],
    "index": 0,
    "kind": "redirect",
    "length": 1,
    "pathname": "/not-found",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
      ],
      "hasSSR": false,
      "hash": "",
      "index": 0,
      "kind": "load",
      "length": 1,
      "params": Object {},
      "pathname": "/first",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "scene": "",
    "search": "",
    "status": 302,
    "url": "/not-found",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {},
  "type": "@@rudy/NOT_FOUND",
}
`;
